### PR TITLE
Edit user profile

### PR DESCRIPTION
**Pull request by anowakowski**
changes for new user profile and user explore improvements with first part of comment functionality

### CODE REVIEW

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.html [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                      <input
+                        matInput
+                        placeholder="edit subtitle"
+                        formControlName="photoSubtitle"
+                      />
+                    </mat-form-field>                  
+                </mat-card-subtitle>
+              </div>
+            </mat-card-header>   
+            <img mat-card-image src="{{ photo.url }}" alt="Photo of a Shiba Inu" />
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.html]: I think we should load data for 'alt' tag dynamically, from data store or simply remove this tag here.

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+/* tslint:disable:no-unused-variable */
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.spec.ts]: [1] Why do we use test components in production code? Is this by convention related to angular coding style?

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      photoDescription: [this.photo.description, Validators.nullValidator]
+    });
+  }
+
+  editPhoto(isEditMode: boolean) {
+    this.isEdit = isEditMode;
+  }
+
+  submitChanges() {
+    console.log('photo changes added');
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.ts]: Is this for testing purposes or production?

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import { Component, OnInit, Input } from '@angular/core';
+import { Photo } from 'src/app/models/photo';
+import { FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { MatDialog } from '@angular/material';
+// tslint:disable-next-line:max-line-length
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.ts]: Is this a special syntax for angular or commented line?

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import { EditUserprofileConfirmationDialogComponent } from '../edit-userprofile-confirmation-dialog/edit-userprofile-confirmation-dialog.component';
+
+@Component({
+  selector: 'app-edit-user-photos',
+  templateUrl: './edit-user-photos.component.html',
+  styleUrls: ['./edit-user-photos.component.scss']
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.ts]: [2] Why styleUrls value is with brackets, whilst the above one not? Is it because of array?

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-userprofile-confirmation-dialog/edit-userprofile-confirmation-dialog.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
+/* tslint:disable:no-unused-variable */
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-userprofile-confirmation-dialog/edit-userprofile-confirmation-dialog.component.spec.ts]: Same like above [1].

File: src/app/modules/pictureapp-explore/components/edit-userProfile/edit-userProfile.component.html [hunk 7]
```diff
@@ -0,0 +1,22 @@
+
+<div layout="row" class="edit-userPfrofile-background-image">
+    <div flex class="edit-userProfile-backgound-image-basePosition"></div>
+    <div class="edit-userProfile-mainPhoto-flex-container-element-image">
+      <img src="{{mainPhoto?.url}}" class="edit-userProfile-mainPhoto-image"/>
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/edit-userProfile/edit-userProfile.component.html]: Is it works same like operator in C# `{{mainPhoto?.url}}`? If so, why do we assume that mainPhoto could be null?

File: src/app/modules/pictureapp-explore/components/edit-userProfile/edit-userProfile.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+/* tslint:disable:no-unused-variable */
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/edit-userProfile/edit-userProfile.component.spec.ts]: Same like above [1].

File: src/app/modules/pictureapp-explore/components/edit-userProfile/edit-userProfile.component.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
+import { Component, OnInit } from '@angular/core';
+import { User } from 'src/app/models/user';
+import { ActivatedRoute } from '@angular/router';
+import { NgxSpinnerService } from 'ngx-spinner';
+import { Photo } from 'src/app/models/photo';
+
+@Component({
+  selector: 'app-edit-user-profile',
+  templateUrl: './edit-userProfile.component.html',
+  styleUrls: ['./edit-userProfile.component.scss']
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/edit-userProfile/edit-userProfile.component.ts]: Same like above [2].

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-userprofile-confirmation-dialog/edit-userprofile-confirmation-dialog.component.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+import { PhotoDetailDialogComponent } from '../../../photo-explore/components/photoDetail-dialog/photoDetail-dialog.component';
+
+export interface DialogData {
+  photo: Photo;
+}
+
+@Component({
+  selector: 'app-edit-userprofile-confirmation-dialog',
+  templateUrl: './edit-userprofile-confirmation-dialog.component.html',
+  styleUrls: ['./edit-userprofile-confirmation-dialog.component.scss']
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-userprofile-confirmation-dialog/edit-userprofile-confirmation-dialog.component.ts]: Same like above [2].

File: src/app/modules/pictureapp-explore/components/nav/nav.component.html [hunk 11]
```diff
@@ -4,8 +4,8 @@
             <mat-icon>menu</mat-icon>
   </button>
   <span>
-      <button mat-button >
-          PictureApp         
+      <button mat-button [routerLink]="['/']" (click)="emitdashboardAsCurrentSection()">
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/nav/nav.component.html]:
```suggestion
      <button mat-button [routerLink]="['/']" (click)="emitDashboardAsCurrentSection()"> 
```
To be followed with convention ;)

File: src/app/modules/pictureapp-explore/components/photo-explore/components/photo-card/photo-card.component.html [hunk 12]
```diff
@@ -1,8 +1,8 @@
-<mat-card class="card">
+<mat-card class="card" [ngClass]="{'photo-card-comment-mode': isCommentMode}">
   <mat-card-header>
     <div
       mat-card-avatar
-      [ngStyle]="{ 'background-image': 'url(' + mainPhoto.url + ')' }"
+      [ngStyle]="{ 'background-image': 'url(' + mainPhoto?.url + ')' }"
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/photo-explore/components/photo-card/photo-card.component.html]: Same like above [3].

File: src/app/modules/pictureapp-explore/components/photo-explore/components/photo-card/photo-card.component.html [hunk 13]
```diff
@@ -13,14 +13,15 @@
...
     </mat-card-title>
-    <mat-card-subtitle>same photo title example</mat-card-subtitle>
+    <mat-card-subtitle>{{currentPhoto.subtitle}}</mat-card-subtitle>
   </mat-card-header>
   <mat-card-content>
     <img
       class="card-img"
       mat-card-image
-      src="{{ currentPhoto.url }}"
+      src="{{ currentPhoto?.url }}"
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/photo-explore/components/photo-card/photo-card.component.html]: Same like above [3].

File: src/app/modules/pictureapp-explore/components/photo-explore/components/photo-card/photo-card.component.html [hunk 14]
```diff
@@ -13,14 +13,15 @@
...
       alt="Photo of a Shiba Inu"
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/photo-explore/components/photo-card/photo-card.component.html]: Same like above [4].

File: src/app/modules/pictureapp-explore/components/photo-explore/components/photo-comment-buttom-sheet/photo-comment-buttom-sheet.component.html [hunk 15]
```diff
@@ -0,0 +1,18 @@
+<div *ngFor="let photoComment of photoComments" class="photo-user-commented-section">
+    <div class="photo-user-commented-section-element"> 
+      <img src="{{photoComment?.user?.photoUrl}}" class="photo-user-commented-image" />
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/photo-explore/components/photo-comment-buttom-sheet/photo-comment-buttom-sheet.component.html]: What kind of image the user will see when `photoComment` will be null?

File: src/app/modules/pictureapp-explore/components/photo-explore/components/photo-comment-buttom-sheet/photo-comment-buttom-sheet.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,28 @@
+/* tslint:disable:no-unused-variable */
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/photo-explore/components/photo-comment-buttom-sheet/photo-comment-buttom-sheet.component.spec.ts]: Same like above [1]

File: src/app/modules/pictureapp-explore/components/photo-explore/components/photo-comment-buttom-sheet/photo-comment-buttom-sheet.component.ts [hunk 17]
```diff
@@ -0,0 +1,23 @@
+import { Component, OnInit, Inject } from '@angular/core';
+import { MatBottomSheetRef, MAT_BOTTOM_SHEET_DATA } from '@angular/material';
+import { Photo } from 'src/app/models/photo';
+import { PhotoComment } from 'src/app/models/photoComment';
+
+@Component({
+  selector: 'app-photo-comment-buttom-sheet',
+  templateUrl: './photo-comment-buttom-sheet.component.html',
+  styleUrls: ['./photo-comment-buttom-sheet.component.scss']
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/photo-explore/components/photo-comment-buttom-sheet/photo-comment-buttom-sheet.component.ts]: Same like above [2]

File: src/app/modules/pictureapp-explore/components/photo-explore/components/photoDetail-dialog/photoDetail-dialog.component.html [hunk 18]
```diff
@@ -2,7 +2,7 @@ <h2 mat-dialog-title>Photo Detail by <b>{{data.user.username}}</b></h2>
 
 <mat-dialog-content>
   <div class="example-container">
-    <img class="img-dialog" src="{{data.currentPhoto.url}}">
+    <img class="img-dialog" src="{{data?.currentPhoto?.url}}">
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/photo-explore/components/photoDetail-dialog/photoDetail-dialog.component.html]: What if data will be null?

File: src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-card/sidenav-userprofile-card.component.html [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+        <mat-card-header >
+          <div 
+            *ngIf="mainPhotoUrl"
+            [ngStyle]="{ 'background-image': 'url(' + mainPhotoUrl + ')' }" 
+            mat-card-avatar 
+            class="sidnav-userprofile-header-image" 
+            [routerLink]="['/editUserProfile']" 
+            (click)="emitEditUserProfileAsCurrentSection()">
+          </div>
+          <mat-card-title class="sidenav-card-header">{{currentUser?.username}}</mat-card-title>
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-card/sidenav-userprofile-card.component.html]: What if `currentUser` will be null?

File: src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-card/sidenav-userprofile-card.component.html [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+          <mat-card-subtitle>
+              {{currentUser?.email}} <br/>
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-card/sidenav-userprofile-card.component.html]: Same like above.

File: src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-card/sidenav-userprofile-card.component.spec.ts [hunk 21]
```diff
@@ -0,0 +1,28 @@
+/* tslint:disable:no-unused-variable */
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-card/sidenav-userprofile-card.component.spec.ts]: Same like above [1].

File: src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-card/sidenav-userprofile-card.component.ts [hunk 22]
```diff
@@ -0,0 +1,34 @@
+import { Component, OnInit, Output, EventEmitter } from '@angular/core';
+import { User } from 'src/app/models/user';
+import { UserService } from 'src/app/services/user.service';
+
+@Component({
+  selector: 'app-sidenav-userprofile-card',
+  templateUrl: './sidenav-userprofile-card.component.html',
+  styleUrls: ['./sidenav-userprofile-card.component.scss']
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-card/sidenav-userprofile-card.component.ts]: Same like above [2].

File: src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-chart/sidenav-userprofile-chart.component.spec.ts [hunk 23]
```diff
@@ -0,0 +1,28 @@
+/* tslint:disable:no-unused-variable */
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-chart/sidenav-userprofile-chart.component.spec.ts]: Same like above [1].

File: src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-chart/sidenav-userprofile-chart.component.ts [hunk 24]
```diff
@@ -0,0 +1,61 @@
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector: 'app-sidenav-userprofile-chart',
+  templateUrl: './sidenav-userprofile-chart.component.html',
+  styleUrls: ['./sidenav-userprofile-chart.component.scss']
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/sidenav/components/sidenav-userprofile-chart/sidenav-userprofile-chart.component.ts]: Same like above [2].

File: src/app/modules/pictureapp-explore/components/sidenav/sidenav.component.ts [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+const SMALL_WIDTH_BREAKPOINT = 720;
+const CURRENT_BTN_COLOR = 'warn';
+const BASIC_BTN_COLOR = 'primary';
+const DASHBOARD_SECTION = 'dashboard';
+const EDIT_USERPROFILE_SECCTION = 'editUserProfile';
+
+@Component({
+  selector: 'app-sidenav',
+  templateUrl: './sidenav.component.html',
+  styleUrls: ['./sidenav.component.scss']
```
Comment (mvein) [inline src/app/modules/pictureapp-explore/components/sidenav/sidenav.component.ts]: Same like above [2].

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.html (hunk 1 above)
Comment (anowakowski) [reply src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.html]: ok removed that.

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.ts (hunk 3 above)
Comment (anowakowski) [reply src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.ts]: its just only test if submitChanges working correctly. Removed that

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.ts (hunk 5 above)
Comment (anowakowski) [reply src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.ts]: its predefined component element its array and you need to use brackets

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.spec.ts (hunk 2 above)
Comment (anowakowski) [reply src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-user-photos/edit-user-photos.component.spec.ts]: its basic ui-test file generating with components. In this file you can check if exists expected function or html generating by self component. The spec file should be close to testing components its typical angular structure

File: src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-userprofile-confirmation-dialog/edit-userprofile-confirmation-dialog.component.spec.ts (hunk 6 above)
Comment (anowakowski) [reply src/app/modules/pictureapp-explore/components/edit-userProfile/components/edit-userprofile-confirmation-dialog/edit-userprofile-confirmation-dialog.component.spec.ts]: Same like reply above

File: src/app/modules/pictureapp-explore/components/edit-userProfile/edit-userProfile.component.html (hunk 7 above)
Comment (anowakowski) [reply src/app/modules/pictureapp-explore/components/edit-userProfile/edit-userProfile.component.html]: the ? opertor use  only for safe view render

File: src/app/modules/pictureapp-explore/components/nav/nav.component.html (hunk 11 above)
Comment (anowakowski) [reply src/app/modules/pictureapp-explore/components/nav/nav.component.html]: ok fixed name convention

**Comment by anowakowski** [comment]
resovle: request#28

**Comment by anowakowski** [comment]
fixed question with:
https://camo.githubusercontent.com/c69711c39c8eb4b2adc283812e5476d44ad49a8c/68747470733a2f2f302e67726176617461722e636f6d2f6176617461722f63613332636539316463336465653138383730333634313531313763316238353f643d68747470732533412532462532466769746875622e6769746875626173736574732e636f6d253246696d6167657325324667726176617461727325324667726176617461722d757365722d3432302e706e6726723d6726733d3430

**Comment by anowakowski** [comment]
fixed question with: 
https://github.com/anowakowski/PictureApp_SPA/pull/28/commits/c3a968c0b430473aae92f62bc4ef53639faffacf
